File: Data Structures & Algorithms/valid-sudoku/submission-0.py
from collections import defaultdict
from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row_map = defaultdict(set)
        col_map = defaultdict(set)
        sqr_map = defaultdict(set)  # key = (row // 3, col // 3)

        for i in range(9):
            for j in range(9):
                val = board[i][j]
                if val == ".":
                    continue

                sqr_index = (i // 3, j // 3)
                
                # Check duplicates
                if (val in row_map[i] or 
                    val in col_map[j] or 
                    val in sqr_map[sqr_index]):
                    return False

                # Add value to sets
                row_map[i].add(val)
                col_map[j].add(val)
                sqr_map[sqr_index].add(val)

        return True

File: Data Structures & Algorithms/surrounded-regions/submission-0.cpp
class Solution {
public:
    int m,n;

    void dfs(int r,int c,vector<vector<char>>& board){
        if(r<0 || r>=m || c<0 || c>=n || board[r][c] != 'O'){
            return;
        }
        board[r][c]='S';
        dfs(r+1,c,board);
        dfs(r-1,c,board);
        dfs(r,c+1,board);
        dfs(r,c-1,board);
    }
    void solve(vector<vector<char>>& board) {
        m=board.size();
        n=board[0].size();

        for(int i=0;i<n;i++){
            dfs(0,i,board);
            dfs(m-1,i,board);
        }

        for(int i=0;i<m;i++){
            dfs(i,0,board);
            dfs(i,n-1,board);
        }

        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(board[i][j]=='O'){
                    board[i][j]='X';
                }
                else if(board[i][j]=='S'){
                    board[i][j]='O';
                }
            }
        }
        
        
    }
};
